clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -547,7 +547,7 @@ public int read(ByteBuffer dst) throws IOException {
                 } else if (unwrapResult.getStatus() == Status.BUFFER_OVERFLOW) {
                     int currentApplicationBufferSize = applicationBufferSize();
                     appReadBuffer = Utils.ensureCapacity(appReadBuffer, currentApplicationBufferSize);
-                    if (appReadBuffer.position() >= currentApplicationBufferSize) {
+                    if (appReadBuffer.position() > currentApplicationBufferSize) {
```

Comment:
To be honest, I'm having a hard time understanding these checks. I guess we are trying to detect when the ssl engine has given us an invalid underflow or overflow status. Presumably if `appReadBuffer` was already large enough, then we shouldn't have gotten the overflow. If that's what we're trying to do, then the original `>=` seems to make sense because it means we didn't actually increase the buffer size? But if the check doesn't actually work, I wonder if we should just remove it.

@rajinisivaram Any insight?

Reply:
@hachikuji We use the buffer sizes from the SSL session. Our SSL implementation assumes that the underlying SSLEngine will not throw overflow/underflow exceptions if we are using matching buffer sizes from the session. In this particular case, we expect that one netReadBuffer can be unwrapped into one appReadBuffer using the buffer sizes from the session. We grow the buffer only if buffer sizes in the session have been updated. If we just removed the check, then in the scenario described in KAFKA-8154, we would end up with an infinite loop. IllegalStateException feels like the better option.

We haven't yet got to the bottom of why this issue was hit in the first place. In particular, the buffer size from the session looks too small and doesn't match any of the JREs I tested with. It will be useful to get more debug level logs. There is more discussion in https://issues.apache.org/jira/browse/KAFKA-8154.

Reply:
Thanks for the background @rajinisivaram. By the way, I noticed one of the checks is defined inconsistently: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java#L344.